OpenRA.Mods.Common/Traits/AttackMove.cs

```diff
@@ -20,56 +23,145 @@ class AttackMoveInfo : ITraitInfo, Requires<IMoveInfo>
+				assaultMoveToken = conditionManager.GrantCondition(self, info.AssaultMoveScanCondition);
+			else if (!assaultActive && assaultMoveToken != ConditionManager.InvalidConditionToken)
+				assaultMoveToken = conditionManager.RevokeCondition(self, assaultMoveToken);
+		}
+
 		public string VoicePhraseForOrder(Actor self, Order order)
 		{
-			if (order.OrderString == "AttackMove")
+			if (order.OrderString == "AttackMove" || order.OrderString == "AssaultMove")
 				return info.Voice;
 
 			return null;
 		}
 
-		void Activate(Actor self)
+		void Activate(Actor self, bool assaultMove)
 		{
+			assaultMoving = assaultMove;
 			self.QueueActivity(new AttackMoveActivity(self, move.MoveTo(TargetLocation.Value, 1)));
 		}
 
 		public void TickIdle(Actor self)
 		{
 			// This might cause the actor to be stuck if the target location is unreachable
 			if (TargetLocation.HasValue && self.Location != TargetLocation.Value)
-				Activate(self);
+				Activate(self, assaultMoving);
 		}
 
 		public void ResolveOrder(Actor self, Order order)
 		{
 			TargetLocation = null;
 
-			if (order.OrderString == "AttackMove")
+			if (order.OrderString == "AttackMove" || order.OrderString == "AssaultMove")
 			{
 				if (!order.Queued)
 					self.CancelActivity();
 
 				TargetLocation = move.NearestMoveableCell(order.TargetLocation);
 				self.SetTargetLine(Target.FromCell(self.World, TargetLocation.Value), Color.Red);
-				Activate(self);
+				Activate(self, order.OrderString == "AssaultMove");
+			}
+		}
+	}
+
+	public class AttackMoveOrderGenerator : UnitOrderGenerator
+	{
+		readonly TraitPair<AttackMove>[] subjects;
+		readonly MouseButton expectedButton;
+
+		public AttackMoveOrderGenerator(IEnumerable<Actor> subjects, MouseButton button)
+		{
+			expectedButton = button;
+
+			this.subjects = subjects.SelectMany(a => a.TraitsImplementing<AttackMove>()
+					.Select(am => new TraitPair<AttackMove>(a, am)))
+				.ToArray();
+		}
+
+		public override IEnumerable<Order> Order(World world, CPos cell, int2 worldPixel, MouseInput mi)
+		{
+			if (mi.Button != expectedButton)
+				world.CancelInputMode();
+
+			return OrderInner(world, cell, mi);
+		}
+
+		protected virtual IEnumerable<Order> OrderInner(World world, CPos cell, MouseInput mi)
+		{
+			if (mi.Button == expectedButton && world.Map.Contains(cell))
+			{
+				world.CancelInputMode();
+
+				var queued = mi.Modifiers.HasModifier(Modifiers.Shift);
+				var orderName = mi.Modifiers.HasModifier(Modifiers.Ctrl) ? "AssaultMove" : "AttackMove";
+				foreach (var s in subjects)
+					yield return new Order(orderName, s.Actor, queued) { TargetLocation = cell };
 			}
 		}
+
+		public override string GetCursor(World world, CPos cell, int2 worldPixel, MouseInput mi)
+		{
+			if (world.Map.Contains(cell))
+				return mi.Modifiers.HasModifier(Modifiers.Ctrl) ? "assaultmove" : "attackmove";
+
+			return "generic-blocked";
```

Comment:
Why `generic-blocked` instead of `assaultmove-blocked`?

Reply:
This was maintaining the existing behaviour - we didn't have any assaultmove-blocked cursors at the time.